velox/functions/prestosql/aggregates/tests/CovarianceAggregationTest.cpp

```diff
@@ -0,0 +1,112 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include "velox/common/base/test_utils/GTestUtils.h"
+#include "velox/exec/tests/utils/PlanBuilder.h"
+#include "velox/functions/prestosql/aggregates/tests/AggregationTestBase.h"
+
+using namespace facebook::velox::exec::test;
+
+namespace facebook::velox::aggregate::test {
+
+class CovarianceAggregationTest
+    : public virtual AggregationTestBase,
+      public testing::WithParamInterface<std::string> {
+ protected:
+  void testGroupBy(const std::string& aggName, const RowVectorPtr& data) {
+    auto partialAgg = fmt::format("{}(c1, c2)", aggName);
+    auto finalAgg = fmt::format("{}(a0)", aggName);
+    auto sql = fmt::format(
+        "SELECT c0, round({}(c1, c2), 2) FROM tmp GROUP BY 1", aggName);
+
+    auto op = PlanBuilder()
+                  .values({data})
+                  .partialAggregation({0}, {partialAgg})
+                  .finalAggregation({0}, {finalAgg})
+                  .project({"c0", "round(a0, cast(2 as integer))"})
+                  .planNode();
+
+    assertQuery(op, sql);
+
+    op = PlanBuilder()
+             .values({data})
+             .singleAggregation({0}, {partialAgg})
+             .project({"c0", "round(a0, cast(2 as integer))"})
+             .planNode();
+
+    assertQuery(op, sql);
+  }
+
+  void testGlobalAgg(const std::string& aggName, const RowVectorPtr& data) {
+    auto partialAgg = fmt::format("{}(c1, c2)", aggName);
+    auto finalAgg = fmt::format("{}(a0)", aggName);
+    auto sql = fmt::format("SELECT round({}(c1, c2), 2) FROM tmp", aggName);
+
+    auto op = PlanBuilder()
+                  .values({data})
+                  .partialAggregation({}, {partialAgg})
+                  .finalAggregation({}, {finalAgg})
+                  .project({"round(a0, cast(2 as integer))"})
+                  .planNode();
+
+    assertQuery(op, sql);
+
+    op = PlanBuilder()
+             .values({data})
+             .singleAggregation({}, {partialAgg})
+             .project({"round(a0, cast(2 as integer))"})
+             .planNode();
+
+    assertQuery(op, sql);
+  }
+};
+
+TEST_P(CovarianceAggregationTest, noNulls) {
+  vector_size_t size = 1'000;
+  auto data = makeRowVector({
+      makeFlatVector<int32_t>(size, [](auto row) { return row % 7; }),
+      makeFlatVector<double>(size, [](auto row) { return row * 0.1; }),
+      makeFlatVector<double>(size, [](auto row) { return row * 0.2; }),
+  });
+
+  createDuckDbTable({data});
+
+  auto aggName = GetParam();
+  testGlobalAgg(aggName, data);
+  testGroupBy(aggName, data);
+}
+
+TEST_P(CovarianceAggregationTest, someNulls) {
```

Comment:
Is it possible to add a test with a nan in the values ?

Reply:
It looks like DuckDB doesn't allow storing NaNs in a table. I'm getting "Out of Range Error: Invalid double value nan" error. @Mytherin Mark, is there a way?

```
terminate called after throwing an instance of 'duckdb::OutOfRangeException'
  what():  Out of Range Error: Invalid double value nan
*** Aborted at 1638561988 (Unix time, try 'date -d @1638561988') ***
*** Signal 6 (SIGABRT) (0x2aca000b812c) received by PID 753964 (pthread TID 0x7ff2af459d80) (linux TID 753964) (maybe from PID 753964, UID 10954) (code: -6), stack trace: ***
    @ 00000000000192fd folly::symbolizer::(anonymous namespace)::innerSignalHandler(int, siginfo_t*, void*)
                       ./folly/experimental/symbolizer/SignalHandler.cpp:449
    @ 00000000000102ab folly::symbolizer::(anonymous namespace)::signalHandler(int, siginfo_t*, void*)
                       ./folly/experimental/symbolizer/SignalHandler.cpp:470
    @ 0000000000000000 (unknown)
    @ 000000000003e530 __GI_raise
    @ 000000000002551c __GI_abort
    @ 00000000000a07d4 __gnu_cxx::__verbose_terminate_handler()
    @ 000000000009e4a5 __cxxabiv1::__terminate(void (*)())
    @ 000000000009e510 std::terminate()
    @ 000000000009e7db __cxa_throw
    @ 000000000101559d duckdb::Value::Value(double)
                       ./velox/external/duckdb/duckdb.cpp:31332
    @ 00000000002b2ab7 duckdb::Value facebook::velox::exec::test::(anonymous namespace)::duckValueAt<(facebook::velox::TypeKind)6>(std::shared_ptr<facebook::velox::BaseVector> const&, int)
                       ./velox/exec/tests/utils/QueryAssertions.cpp:54
    @ 000000000029bde9 facebook::velox::exec::test::DuckDbQueryRunner::createTable(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::vector<std::shared_ptr<facebook::velox::RowVector>, std::allocator<std::shared_ptr<facebook::velox::RowVector> > > const&)::$_0::operator()() const
                       ./velox/exec/tests/utils/QueryAssertions.cpp:397
```

Reply:
DuckDB does not currently allow NaNs into the system, they are instead converted to `NULL` values when ingested or an error is thrown when they are generated from a function call.